### PR TITLE
Travisci

### PR DESCRIPTION
**Pull request by pleimer**
Implement Travis CI to run ansible tests

### CODE REVIEW

**Comment by leifmadsen** [comment]
Looking through the logs here, the error is here:

```
    TASK [smartgateway : Write out configmap for smart gateway] ********************
    fatal: [localhost]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'meta' is undefined"}
```

For some reason the `meta` map isn't getting through to the molecule setup. Will need to try and reproduce it locally and determine why it's not seeing it. What is needed is basically a couple of values from `meta.namespace` and another value I can't remember. The test fails because that `meta` map isn't populated (I assume).

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+services:
+  - docker
+before_install:
+  - pip install --upgrade pip --user
+  - pip install molecule docker --user
+script:
+  - molecule --version
+  - ansible --version
+  - molecule test
+  #- docker build --tag redhat-service-assurance/smart-gateway-operator:latest .
```
Comment (leifmadsen) [inline .travis.yml]: You can just delete this line and the one below it since we're going to automatically build and deploy on merges to master.

**Comment by leifmadsen** [comment]
So @pleimer I was able to easily reproduce this locally.

```
git clone <this repo> ; cd <this repo>
virtualenv .venv
source .venv/bin/activate
pip install ansible molecule docker
molecule test
```

I started going through the molecule stuff, but basically I haven't read any documentation, so I don't really fully understand the structure.

What I did see was that the `prepare.yaml` doesn't seem to fire for some reason. If it did, we should see a custom resource being made, which loads the `_cr.yaml` from the `crds/` directory (the `_crd.yaml` contains the custom resource definition, which tells you what CR (custom resource) is valid). The reason that's important is that it contains the `metadata` that should get passed into the Operator, so that the role has some information to work on. You can see the incoming values from the custom resource being represented in the `roles/` directory within some templates (for example).

In the Ansible operator, it'll look like `{{ meta.name }}` or whatever.

At this point, just need to learn molecule. I look forward to your happy hour presentation about how molecule works :) How you solve this problem will be an interesting case study.

File: .travis.yml (hunk 1 above)
Comment (pleimer) [reply .travis.yml]: 'molecule test' runs the playbooks, are you referring to the commented and '--version' lines?

Comment (leifmadsen) [reply .travis.yml]: Not above... below :)

```
#- docker build --tag redhat-service-assurance/smart-gateway-operator:latest .
#- docker images
```

**Comment by pleimer** [comment]
Embarrassingly it looks like I was running the wrong tests - it is running default instead of running test local. However there are other issues with these tests

File: .travis.yml [hunk 2]
```diff
@@ -6,10 +6,9 @@ services:
   - docker
 before_install:
   - pip install --upgrade pip --user
-  - pip install molecule docker --user
+  - pip install molecule docker kubernetes openshift --user
```
Comment (leifmadsen) [inline .travis.yml]: Add `jmespath` as well here

**Comment by leifmadsen** [comment]
> Embarrassingly it looks like I was running the wrong tests - it is running default instead of running test local. However there are other issues with these tests

That actually helped a bunch. I see it creating the `CR` now. I'm attempting to run locally. I had it going before, but it hung on a Gathering Metadata step. I have killed it and am now trying again.

**Comment by pleimer** [comment]
> > Embarrassingly it looks like I was running the wrong tests - it is running default instead of running test local. However there are other issues with these tests
> 
> That actually helped a bunch. I see it creating the `CR` now. I'm attempting to run locally. I had it going before, but it hung on a Gathering Metadata step. I have killed it and am now trying again.

Any success?

**Comment by leifmadsen** [comment]
> > > Embarrassingly it looks like I was running the wrong tests - it is running default instead of running test local. However there are other issues with these tests
> > 
> > 
> > That actually helped a bunch. I see it creating the `CR` now. I'm attempting to run locally. I had it going before, but it hung on a Gathering Metadata step. I have killed it and am now trying again.
> 
> Any success?

So it makes it much further, but it still fails for me.

```
--> Action: 'converge'
    
    PLAY [Build Operator in Kubernetes docker container] ***************************
    
    TASK [Gathering Facts] *********************************************************
    ok: [kind-test-local]
    
    TASK [Get existing image hash] *************************************************
    ok: [kind-test-local]
    
    TASK [Build Operator Image] ****************************************************
    changed: [kind-test-local]
    
    PLAY [Converge] ****************************************************************
    
    TASK [Gathering Facts] *********************************************************
    ok: [localhost]
    
    TASK [Delete the Operator Deployment] ******************************************
    ok: [localhost]
    
    TASK [Wait 30s for Operator Deployment to terminate] ***************************
    skipping: [localhost]
    
    TASK [Create the Operator Deployment] ******************************************
    changed: [localhost]
    
    TASK [Create the smartgateway.infra.watch/v1alpha1.SmartGateway] ***************
    changed: [localhost]
    
    TASK [debug] *******************************************************************
    ok: [localhost] => {
        "msg": "smartgateway.infra.watch/v1alpha1 -- SmartGateway -- osdk-test -- white"
    }
    
    TASK [Wait 40s for reconciliation to run] **************************************
    FAILED - RETRYING: Wait 40s for reconciliation to run (10 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (9 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (8 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (7 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (6 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (5 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (4 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (3 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (2 retries left).
    FAILED - RETRYING: Wait 40s for reconciliation to run (1 retries left).
    fatal: [localhost]: FAILED! => {"attempts": 10, "changed": false, "resources": [{"apiVersion": "smartgateway.infra.watch/v1alpha1", "kind": "SmartGateway", "metadata": {"creationTimestamp": "2019-06-22T12:37:56Z", "generation": 1, "name": "white", "namespace": "osdk-test", "resourceVersion": "694", "selfLink": "/apis/smartgateway.infra.watch/v1alpha1/namespaces/osdk-test/smartgateways/white", "uid": "8f61a48a-94ea-11e9-8133-0242e5cfaa83"}, "spec": {"amqpmetric_url": "qdr-white-normal.osdk-test.svc.cluster.local:5672/collectd/telemetry", "debug": "false", "size": 1}}]}
    
    TASK [debug cr] ****************************************************************
    ok: [localhost] => {
        "debug_cr": {
            "apiVersion": "smartgateway.infra.watch/v1alpha1",
            "kind": "SmartGateway",
            "metadata": {
                "creationTimestamp": "2019-06-22T12:37:56Z",
                "generation": 1,
                "name": "white",
                "namespace": "osdk-test",
                "resourceVersion": "694",
                "selfLink": "/apis/smartgateway.infra.watch/v1alpha1/namespaces/osdk-test/smartgateways/white",
                "uid": "8f61a48a-94ea-11e9-8133-0242e5cfaa83"
            },
            "spec": {
                "amqpmetric_url": "qdr-white-normal.osdk-test.svc.cluster.local:5672/collectd/telemetry",
                "debug": "false",
                "size": 1
            }
        },
        "failed_when_result": false
    }
    
    TASK [debug memcached lookup] **************************************************
    ok: [localhost] => {
        "deploy": [],
        "failed_when_result": false
    }
    
    TASK [get operator logs] *******************************************************
    ok: [localhost]
    
    TASK [debug] *******************************************************************
    ok: [localhost] => {
        "log.stdout_lines": "VARIABLE IS NOT DEFINED!"
    }
    
    TASK [fail] ********************************************************************
    fatal: [localhost]: FAILED! => {"changed": false, "msg": "Failed on action: converge"}
    
    PLAY RECAP *********************************************************************
    kind-test-local            : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
    localhost                  : ok=9    changed=2    unreachable=0    failed=1    skipped=1    rescued=1    ignored=0
```

**Comment by leifmadsen** [comment]
Looked at this again for a couple minutes. I think the issue is that we need to start an AMQ Interconnect so that the SG can actually connect to something on boot. This will need to work similarly to how the SG unit tests work (which is nice because we'll have a similar setup method).

**Comment by leifmadsen** [comment]
I'm just going to close this since it's so far out of date and our CI work has changed fairly significantly over the last little bit, that I'm not sure this is a valid base to continue developing from. Thanks!

**Pull request closed without merging**